ldap/servers/slapd/back-ldbm/db-mdb/mdb_import_threads.c

```diff
@@ -387,44 +633,323 @@ dbmdb_import_add_created_attrs(Slapi_Entry *e)
+    }
+
+    dnrc = get_entry_type(wqelmt, &param->sdn);
+    if (dnrc == DNRC_SUFFIX) {
+        if ( param->eid != 1) {
+            dnrc = DNRC_BAD_SUFFIX_ID;
+        } else {
+            rdn = slapi_sdn_get_dn(&param->sdn);
+            nrdn = slapi_sdn_get_ndn(&param->sdn);
+            if (!(param->flags & EIP_RDN)) {
+                param->ekey.mv_data = (void*)nrdn;
+                param->ekey.mv_size = strlen(nrdn)+1;
+                param->pkey.mv_data = NULL;
+                param->pkey.mv_size = 0;
+            }
+        }
+    } else if (dnrc == DNRC_OK) {
+        if (param->flags & EIP_RDN) {
+            if (!pidfromkey) {
+                dnrc = DNRC_NOPARENT_ID;
+            }
+            rdn = slapi_sdn_get_dn(&param->sdn);
+            nrdn = slapi_sdn_get_ndn(&param->sdn);
+        } else {
+            slapi_sdn_get_rdn(&param->sdn, &srdn);
+            rdn = slapi_rdn_get_rdn(&srdn);
+            nrdn = slapi_rdn_get_nrdn(&srdn);
+            /* Check for DN validity */
+            if (nrdn == NULL) {
+                /* ldap_explode failed */
+                dnrc = DNRC_BADDN;
+            } else {
+                param->ekey.mv_data = (void*)slapi_sdn_get_ndn(&param->sdn);
+                param->ekey.mv_size = strlen(param->ekey.mv_data)+1;
+                param->pkey.mv_data = (void*)slapi_dn_find_parent_ext(param->ekey.mv_data, 0);
+                param->pkey.mv_size = strlen(param->pkey.mv_data)+1;
+                if (!param->pkey.mv_data) {
+                    dnrc = DNRC_NOPARENT_DN;
+                }
+            }
+        }
+    }
+    if (dnrc == DNRC_OK) {
+        /* Regular entry should have a parent info in the private db */
+        /* lets lookup for the parent ndn */
+        rc = dbmdb_privdb_get(param->db, 0, &param->pkey, &data);
+        if (rc == MDB_NOTFOUND) {
+            if (param->flags & EIP_WAIT) {
+                dnrc = DNRC_WAIT;
+            } else {
+                dnrc = DNRC_NOPARENT_ID;
+            }
+        } else if (rc) {
+            dnrc = DNRC_ERROR;
+        } else {
+            /* Everything is ok, lets duplicate the parent info for the worker thread  */
+            wqelmt->parent_info = (ID*)slapi_ch_calloc(ALIGN_TO_ID_SLOT(data.mv_size), sizeof(ID));
+            memcpy(wqelmt->parent_info, data.mv_data, data.mv_size);
+            pinfo = wqelmt->parent_info;
+        }
+    }
+    if (dnrc == DNRC_SUFFIX || dnrc == DNRC_OK) {
+        /* Note: If someday we needs to rebuild the entry dn, we could store it in the entry info
+         *  and build it from parent entry info + entry rdn. But as of today we do not need it.
+         */
+        len = strlen(nrdn) + strlen(rdn) + 2 + (5 + pinfo[1]) * sizeof(ID);
+        data.mv_data = wqelmt->entry_info = (ID*)slapi_ch_calloc(ALIGN_TO_ID_SLOT(len), sizeof(ID));
+        data.mv_size = len;
+        wqelmt->entry_info[0] = param->eid;
+        wqelmt->entry_info[1] = pinfo[1] + 1;
```

Comment:
why parentID + 1 ?

Reply:
parentID is pinfo[0], pinfo[1] is the number of ancestors which is 1 + number of parent ancestors